tests/schema/tests.py

```diff
@@ -2481,6 +2488,327 @@ def test_text_field_with_db_index(self):
+        self.assertIn(index.name, self.get_constraints(table))
+        self.assertIs(sql.references_column(table, 'weight'), True)
+        # Remove index.
+        with connection.schema_editor() as editor:
+            editor.remove_index(Author, index)
+        self.assertNotIn(index.name, self.get_constraints(table))
+
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_func_index_without_name(self):
+        with self.assertRaisesMessage(ValueError, 'Index.name needs to be set when passed expressions.'):
+            Index(Lower('title').desc())
+
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_func_index_field_and_expression(self):
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(Book)
+        index = Index(
+            F('author').desc(), Lower('title').asc(), 'pub_date',
+            name='composite_mixed_func_idx'
+        )
+        # Add index.
+        with connection.schema_editor() as editor:
+            editor.add_index(Book, index)
+            sql = index.create_sql(Book, editor)
+        table = Book._meta.db_table
+        constraints = self.get_constraints(table)
+        # SQLite does not support introspection of expressions in indexes
+        # https://www.sqlite.org/pragma.html#pragma_index_xinfo
+        if connection.features.supports_index_column_ordering and not connection.vendor == 'sqlite':
+            self.assertIndexOrder(table, index.name, ['DESC', 'ASC', 'ASC'])
+        self.assertEqual(len(constraints[index.name]['columns']), 3)
+        self.assertEqual(constraints[index.name]['columns'][0], 'author_id')
+        self.assertEqual(constraints[index.name]['columns'][-1], 'pub_date')
+        # SQL contains database functions and columns.
+        self.assertIs(sql.references_column(table, 'author_id'), True)
+        self.assertIs(sql.references_column(table, 'title'), True)
+        self.assertIs(sql.references_column(table, 'pub_date'), True)
+        self.assertIn('LOWER(%s)' % editor.quote_name('title'), str(sql))
+        # Remove index.
+        with connection.schema_editor() as editor:
+            editor.remove_index(Book, index)
+        self.assertNotIn(index.name, self.get_constraints(table))
+
+    @skipUnlessDBFeature('supports_expression_indexes', 'supports_index_collations')
+    def test_func_index_with_custom_collation(self):
+        collation = connection.features.test_collations.get('non_default')
+        if not collation:
+            self.skipTest(
+                'This backend does not support case-insensitive collations.'
+            )
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+            editor.create_model(BookWithSlug)
+        index = Index(
+            Collate(F('title'), collation=collation).desc(),
+            Collate(Cast('author', TextField()).desc(), collation=collation),
```

Comment:
Indexes on `TextField`s are not supported on Oracle (see `supports_index_on_text_field` feature flag), that's why it fails. Including a `TextField` in this index is not crucial so I will remove it.

Reply:
Ahh, sorry we cast it to the `NVARCHAR2(2000)` so that's not the issue. Nevertheless, it fails on my local instance with:
```
django.db.utils.DatabaseError: ORA-01450: maximum key length (6398) exceeded
```
because I have a 8K block size. I will remove `TextField` to avoid issues on such configurations.

Reply:
Also this test is a bit redundant with `test_index_with_collate_f_expressions`. I think we can remove it.